### DISCUSSION TITLE
How to get widgets list and macros list from JS?

### DISCUSSION THREAD

**Discussion by linonetwo** [TiddlyWiki/TiddlyWiki5#6706] [Q&A] [answered]
I'm trying to autocomplete the available widgets and fields when typing `<$`, and autocomplete `<<`'s available macros.

I didn't find a place in $tw.wiki where I can get a full list of widgets and macros.

Seems I have had too many questions since I started writing the WYSIWYG editor...

**Comment by linonetwo** [comment]
### Macros

Seems it didn't store the macro in a list or global variable when encountering it.

https://github.com/Jermolene/TiddlyWiki5/blob/696c5c9c7a512cb2b146b2e4df129c20cebb6515/core/modules/widgets/importvariables.js#L59-L74

And when parsing macros, they are transformed into `<$set` widget in place

https://github.com/Jermolene/TiddlyWiki5/blob/696c5c9c7a512cb2b146b2e4df129c20cebb6515/core/modules/parsers/wikiparser/rules/macrodef.js#L80-L89

So maybe there is no global macro list, macros are re-parsed every time and put into the tree when generating every widget tree.

**Comment by saqimtiaz** [comment, marked as answer]
For querying available widgets, see the discussion at this issue: https://github.com/Jermolene/TiddlyWiki5/issues/6453

Macros are trickier since they can be defined by users and are just variables. You could try autocompleting based on the variables defined in the context of the editor.
https://tiddlywiki.com/#variables%20Operator

**Reply by linonetwo** [reply]
Thanks, I think `[variables[]]` will work for macros, since every variable can be used in the `<< >>`.

**Reply by linonetwo** [reply]
About widgets, seems you don't have a ready-to-use implementation in that issue?

**Reply by linonetwo** [reply]
Why my `$tw.wiki.compileFilter('[variables[]]')()` always returns `[]`?

---

Oh, I need to pass my widget into it, otherwise, it will use rootWidget which don't have any variables.


`$tw.wiki.compileFilter('[variables[]]')(undefined, myWidget)`

<img width="1753" alt="截屏2022-05-23 23 39 21" src="https://user-images.githubusercontent.com/3746270/169856327-44ad0a85-aa64-4f93-830d-8e4b4363638b.png">

**Reply by linonetwo** [reply]
Widgets are in `$tw.rootWidget.widgetClasses`, you can use this to create a widget filter. But I don't know if this is the complete set of widgets.

<img width="981" alt="截屏2022-05-24 21 10 54" src="https://user-images.githubusercontent.com/3746270/170042885-ad177e16-1825-4938-88ac-dedb00251a94.png">

**Reply by saqimtiaz** [reply]
That will work, the implementation should be the same as for the deserializers[] operator: https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/filters/deserializers.js

Note that you can just have this filter as part of your plugin until there is core support.

**Comment by linonetwo** [comment]
I can get macro list now, but I hope there can be a way to get corresponding examples...

Currently, user will still don't know what these meacros means.

![截屏2022-05-23 23 46 31](https://user-images.githubusercontent.com/3746270/169857781-343c7823-d11a-42a2-8a29-277d2a9abeb8.png)

Like in https://github.com/Jermolene/TiddlyWiki5/discussions/6670 , I hope there can be a way to add fields for example usages to macros and widgets. Or probably use a new system tag to match the readme of macros to the macro definitions. And show readme in the preview of dropdown just like what we have in snippet autocomplete dropdown and wikilink autocomplete dropdown:


<img width="785" alt="截屏2022-05-23 23 51 03" src="https://user-images.githubusercontent.com/3746270/169858449-cad5c65b-d3c3-40db-94c5-857f218bdd86.png">
